modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,454 @@
+		 *
+		 * By default this will be empty array. Consumer code will need to use this filter in
+		 * order to return the list of items which needs to be processed in current batch.
+		 *
+		 * @param array  $items  Items for the current batch.
+		 * @param int    $job_id Background job ID.
+		 * @param string $name   Job identifier or name.
+		 * @param array  $data   Job data.
+		 *
+		 * @since n.e.x.t
+		 */
+		$items = apply_filters( 'perflab_job_batch_items', $items, $this->job_id, $this->name, $this->data );
+
+		return (array) $items;
+	}
+
+	/**
+	 * Process the batch item.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $item Batch item for the job.
+	 *
+	 * @return void
+	 */
+	public function process( $item ) {
+		/**
+		 * Hook to this action to run the job.
+		 *
+		 * Concrete implementations would add the logic to run this job.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param mixed $item Batch item for the job.
+		 * @param int   $job_id Job ID.
+		 * @param array $data   Job data.
+		 */
+		do_action( 'perflab_process_' . $this->name . '_job_item', $item, $this->job_id, $this->data );
+	}
+
+	/**
+	 * Get number of retries attempted for a job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int
+	 */
+	public function get_retries() {
+		$retries = get_term_meta( $this->job_id, self::JOB_RETRY_META_KEY, true );
+		$retries = absint( $retries );
+
+		return $retries;
+	}
+
+	/**
+	 * Locks the process. It tells that process is running.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	public function lock() {
+		$time = time();
+
+		update_term_meta( $this->job->job_id, self::JOB_LOCK_META_KEY, $time );
+		$this->job->set_status( self::JOB_STATUS_RUNNING );
+	}
+
+	/**
+	 * Unlocks the process.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	public function unlock() {
+		delete_term_meta( $this->job->job_id, self::JOB_LOCK_META_KEY );
+	}
+
+	/**
+	 * Records the error for the current process run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error Error instance.
+	 *
+	 * @todo Implement the error recording activity.
+	 *
+	 * @return void
+	 */
+	public function record_error( WP_Error $error ) {
+		$job_failure_data = $error->get_error_data( 'perflab_job_failure' );
+
+		if ( ! empty( $job_failure_data ) ) {
+			$this->job->set_status( self::JOB_STATUS_FAILED );
+
+			update_term_meta( $this->job_id, self::JOB_ERRORS_META_KEY, $job_failure_data );
+			update_term_meta( $this->job_id, self::JOB_RETRY_META_KEY, ( $this->get_retries() + 1 ) );
+		}
+	}
+
+	/**
+	 * Return max number of retries to attempt for a failed job to re-run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return int
+	 */
+	private function max_retries_allowed() {
+		/**
+		 * Number of attempts to try to re-run a failed job.
+		 * Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $retry Number of retries allowed for a job to re-run.
+		 *
+		 * @return int
+		 */
+		return apply_filters( 'perflab_job_max_retries_allowed', 3 );
+	}
```

Comment:
We don't need separate function for `apply_filters` you can add it in `should_run()`.